A68-random-subsetting.md

```diff
@@ -0,0 +1,141 @@
+![State Diagram](A68_graphics/subsetting500-10-5.png)
+* N_clients = 2000, N_servers = 10, subset_size = 5
+![State Diagram](A68_graphics/subsetting2000-10-5.png)
+
+#### Low connection churn during server rollouts
+
+The graphs provided in the previous section prove this is the case in practice (we rollout all servers in the middle of every test, and there is no visible increase in the number of connections per server) Low connection churn during server rollouts is the primary motivation why rendezvous hashing was used as the subsetting algorithm: it guaranties that if a single server is either added or removed to the endpoints list, every client will update at most 1 entry in its subset. This is because the hashes for all unaffected servers remain the same, which guarantees that the order of the servers after sorting also remains stable. The same logic applies to the situation when multiple servers got updated. 
+
+### LB Policy Config and Parameters
+
+The `random_subsetting` LB policy config will be as follows.
+
+```proto
+message LoadBalancingConfig {
+    oneof policy {
+        RandomSubsettingLbConfig random_subsetting = 21 [json_name = "random_subsetting"];
+    }
+}
+message RandomSubsettingLbConfig {
+    // subset_size indicates how many backends every client will be connected to.
+    // Default is 20.
+    google.protobuf.UInt32Value subset_size = 1;
+
+    // The config for the child policy.
+    repeated LoadBalancingConfig child_policy = 2;
+}
+```
+
+
+### Handling Parent/Resolver Updates
+
+When the resolver updates the list of endpoints, or the LB config changes, Random subsetting LB will run the subsetting algorithm, described above, to filter the endpoint list. Then it will create a new resolver state with the filtered list of the endpoints and pass it to the child LB. Attributes and service config from the old resolver state will be copied to the new one. 
+
+### Handling Subchannel Connectivity State Notifications
+
+Random subsetting LB will simply redirect all requests to the child LB without doing any additional processing. This also applies to all other callbacks in the LB interface, besides the one that handles resolver and config updates (which is described in the previous section). This is possible because random subsetting LB doesn't store or manage sub-connections - it acts as a simple filter on the resolver state, and that's why it can redirect all actual work to the child LB. 
+
+### xDS Integration
+
+Random subsetting LB won't depend on xDS in any way. People may choose to initialize it by directly providing service config. We will only provide a corresponding xDS policy wrapper to allow configuring this LB via xDS.
+
+#### Changes to xDS API
+
+`random_subsetting` will be added as a new LB policy.
+
+```proto
+package envoy.extensions.load_balancing_policies.random_subsetting.v3;
+message RandomSubsetting {
+  google.protobuf.UInt32Value subset_size = 1;
+  repeated LoadBalancingConfig child_policy = 2;
```

Comment:
I think this will want to be:

```
config.cluster.v3.LoadBalancingPolicy child_policy = 2;
```

Note that it does *not* need to be `repeated`, because in xDS the list of options is inside of the `LoadBalancingPolicy` message.

Reply:
This was a bad copy-paste. Updated.  
Should the corresponding field in the LB policy config be repeated? I removed repeated keyword for now as in this case the child policy is mandatory.

Reply:
The reason for `repeated` here is to allow new policies to be configured without breaking older clients.  The client will iterate through the list of child policies, stopping at the first one that it supports.  So let's say that you start with a bunch of clients that support only policy A, and now you want to move to a new policy B, but not all of your clients support B yet.  The control plane can send the list [B, A], which will allow new clients to use B while still allowing older clients to use A.

In the gRPC service config, the `LoadBalancingConfig` message represents a single policy, so we need to make the field `repeated` wherever it gets used.  In xDS, the `config.cluster.v3.LoadBalancingPolicy` message itself contains [a `repeated` list of policies](https://github.com/envoyproxy/envoy/blob/7f7234be42dd7f90c612c193187a65775c119607/api/envoy/config/cluster/v3/cluster.proto#L1235), so there's no need to make it repeated.  (In hindsight, I wish we'd made the same kind of intermediate message in the gRPC service config, so that we didn't always have to remember to make the field `repeated` wherever we use it.  At the time we were concerned with trying to minimize the length of the JSON representation to allow it to squeeze into DNS TXT records, which is why we chose the current approach.)

So yes, please do add `repeated` in the gRPC service config representation.

Reply:
Added `repeated` back to RandomSubsettingLbConfig.